### ISSUE TITLE
RA doesn't seem to respect cfg on fields

### ISSUE THREAD

**Issue by Lokathor** [rust-analyzer/rust-analyzer#4063]
```rust
pub struct ZlibByteIter<'s> {
  pub(crate) current: &'s [u8],
  pub(crate) idat_iter: IdatIter<'s>,
  #[cfg(not(feature = "png_skip_adler_checks"))]
  pub(crate) adler: u32,
}

impl<'s> ZlibByteIter<'s> {
  fn from_current_and_iter(current: &'s [u8], idat_iter: IdatIter<'s>) -> Self {
    #[cfg(not(feature = "png_skip_adler_checks"))]
    {
      Self { current, idat_iter, adler: 1 }
    }
    #[cfg(feature = "png_skip_adler_checks")]
    {
      Self { current, idat_iter }
    }
  }
}
```

This will report the second cfg variant as missing the `adler` field, which doesn't exist under that configuration.

**Comment by bjorn3** [comment]
I think this should have been fixed by https://github.com/rust-analyzer/rust-analyzer/pull/3880, but that PR was merged 11 days ago.

**Comment by Lokathor** [comment]
It seems I have, `rust-analyzer version: 2020-04-13 (c388130)`, which is newer than 11 days ago if my math is right

**Comment by edwin0cheng** [comment]
> which doesn't exist under that configuration.

What's your `rust-analyzer.cargo.features` settings ?

**Comment by flodiebold** [comment]
I think the problem might rather be that we don't support `cfg` on *blocks* (or do we?) :thinking:

**Comment by bnjjj** [comment]
For now `cfg` is partially supported on struct fields, and function as far as I remember. And when I say partially it's mainly because it just check the default features in your cargo.toml. I think it should work with this kind of syntax:
```rust
pub struct ZlibByteIter<'s> {
  pub(crate) current: &'s [u8],
  pub(crate) idat_iter: IdatIter<'s>,
  #[cfg(not(feature = "png_skip_adler_checks"))]
  pub(crate) adler: u32,
}

impl<'s> ZlibByteIter<'s> {
  fn from_current_and_iter(current: &'s [u8], idat_iter: IdatIter<'s>) -> Self {
      Self {
          current,
          idat_iter,
          #[cfg(not(feature = "png_skip_adler_checks"))]
          adler: 1
     }
  }
}
```

Or you can also have 2 methods `fn from_current_and_iter(current: &'s [u8], idat_iter: IdatIter<'s>) -> Self` on with the feature flag enabled and another one without. Because for now we don't support `cfg` attributes on `block`.

I will try to implement it in the next days :)

**Comment by lnicola** [comment]
Closed by #4063.
